java-frontend/src/main/java/org/sonar/java/resolve/LeastUpperBound.java

```diff
@@ -0,0 +1,315 @@
+      JavaType subs2 = typeSubstitution2.substitutedType(typeVar);
+
+      JavaType newSubs = getNewTypeArgumentType(subs1, subs2);
+      newTypeSubstitution.add(typeVar, newSubs);
+    }
+
+    return parametrizedTypeCache.getParametrizedTypeType(type1.symbol, newTypeSubstitution);
+  }
+
+  private JavaType getNewTypeArgumentType(JavaType type1, JavaType type2) {
+    boolean isWildcard1 = type1.isTagged(JavaType.WILDCARD);
+    boolean isWildcard2 = type2.isTagged(JavaType.WILDCARD);
+
+    JavaType result;
+    if (type1.equals(type2)) {
+      result = type1;
+    } else if (isWildcard1 && isWildcard2) {
+      result = lctaBothWildcards((WildCardType) type1, (WildCardType) type2);
+    } else if (isWildcard1 ^ isWildcard2) {
+      JavaType rawType = isWildcard1 ? type2 : type1;
+      WildCardType wildcardType = (WildCardType) (isWildcard1 ? type1 : type2);
+      result = lctaOneWildcard(rawType, wildcardType);
+    } else {
+      result = lctaNoWildcard(type1, type2);
+    }
+    return result;
+  }
+
+  private JavaType lctaOneWildcard(JavaType rawType, WildCardType wildcardType) {
+    if (wildcardType.boundType == WildCardType.BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Lists.newArrayList(rawType, wildcardType.bound));
+      return parametrizedTypeCache.getWildcardType(glb, WildCardType.BoundType.SUPER);
+    }
+    JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(rawType, wildcardType.bound));
+    return parametrizedTypeCache.getWildcardType(lub, WildCardType.BoundType.EXTENDS);
+  }
+
+  private JavaType lctaBothWildcards(WildCardType type1, WildCardType type2) {
+    if (type1.boundType == WildCardType.BoundType.SUPER && type2.boundType == WildCardType.BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Lists.newArrayList(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(glb, WildCardType.BoundType.SUPER);
+    }
+    if (type1.boundType == WildCardType.BoundType.EXTENDS && type2.boundType == WildCardType.BoundType.EXTENDS) {
+      JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(lub, WildCardType.BoundType.EXTENDS);
+    }
+    if (type1.bound.equals(type2.bound)) {
+      return type1.bound;
+    }
+    return symbols.unboundedWildcard;
+  }
+
+  private JavaType lctaNoWildcard(JavaType type1, JavaType type2) {
+    JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(type1, type2));
+    return parametrizedTypeCache.getWildcardType(lub, WildCardType.BoundType.EXTENDS);
+  }
+
+  /**
+   * From JLS 8 5.1.10 - greatest lower bound :  glb(V1,...,Vm) is defined as V1 & ... & Vm.
+   * @param types
+   * @return
+   */
+  private static Type greatestLowerBound(List<Type> types) {
+    // TODO SONARJAVA-1632 implement, should return intersection of all types, not only first one
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)